src/net/gosha/atproto/client.clj

```diff
@@ -1,39 +1,98 @@
+  env-keys
+  {:openapi-spec "ATPROTO_OPENAPI_SPEC"
+   :base-url "ATPROTO_BASE_URL"
+   :username "ATPROTO_USERNAME"
+   :app-password "ATPROTO_APP_PASSWORD"})
+
+(defn- build-config
+  "Build a configuration map based on defaults, env vars and provided values"
+  [& {:as user-config}]
+  (-> {:openapi-spec "./resources/atproto-xrpc-openapi.2024-12-18.json"}
+    (into (map (fn [[cfg-key env-var]]
+                 (when-let [val (System/getenv env-var)]
+                   {cfg-key val}))
+            env-keys))
+    (merge user-config)))
+
+(defn- add-authentication-header [token]
   {:name ::add-authentication-header
    :enter (fn [ctx]
-            (if token
-              (assoc-in ctx [:request :headers "Authorization"] (str "Bearer " token))
-              ctx))})
+            (assoc-in ctx [:request :headers "Authorization"]
+              (str "Bearer " token)))})
+
+;; TODO: automatically refresh expired tokens
+(defn authenticate
+  "Given an api session, authenticate with the atproto API using an app password
+  and return an authenticated api session.
+
+  Note that the session will only work as long as the token is valid. If the
+  token expires, authenticate will need to be called again."
+  [session]
+  (let [{:keys [username app-password openapi-spec base-url]} (::config session)
+        response @(martian/response-for
+                    session
+                    :com.atproto.server.create-session
+                    {:identifier username :password app-password})
+        token (get-in response [:body :accessJwt])]
+    (when-not token
+      (if (:status response)
+        (throw (ex-info (format "Authorization failed (%s)" (:status response))
+                 {:response response}))
+        (throw (ex-info "Authorization failed (unknown error)"
+                 {:response response}))))
+    (martian-http/bootstrap-openapi openapi-spec
+      {:server-url base-url
+       :interceptors (cons (add-authentication-header token)
+                       martian-http/default-interceptors)})))
 
-(def api (martian-http/bootstrap-openapi
-           openapi-url
-           {:server-url (:base-url @core/config)
-            :interceptors (concat [(add-authentication-header (:auth-token @core/config))]
-                                  martian-http/default-interceptors)}))
+(defn init
+  "Create and return a new api session. Valid configuration options are:
+
+  :username - (optional) Bluesky username.
+  :app-password - (optional) Bluesky appplication-specific password.
+  :base-url - Bluesky endpoint. Mandatory. Use 'https://public.api.bsky.app'
+              for unauthenticated access, 'https://bsky.social' for
+              authenticated, or a different endpoint if you know what you're
+              doing.
+  :openapi-spec - OpenAPI JSON specification. Defaults to the included spec.
+
+  All options can be specifed via an environment variable prefixed by `ATPROTO_`
+  (e.g. ATPROTO_BASE_URL)."
+  [& {:as options}]
```

Comment:
I didn't know this construct — it means I can call the function with keyword arguments without enclosing them into a map? That's really nice